src/apps/dictionaries/pages/AddBulkConceptsPage.tsx

```diff
@@ -22,22 +40,147 @@ const useStyles = makeStyles((theme: Theme) =>
+    },
+    sourcesDropdownHeader: {
+      padding: "0.5rem 1rem"
+    },
+    input: {
+      cursor: "pointer",
+      borderBottom: "1px dotted black",
+      paddingBottom: "0.25rem"
+    },
+    underline: {
+      "&&&:before": {
+        borderBottom: "none"
+      },
+      "&&:after": {
+        borderBottom: "none"
+      }
+    },
+    sourceIcon: {
+      marginRight: "0.2rem",
+      fill: "#8080809c"
+    },
+    searchInput: {
+      textAlign: "center",
+      fontSize: "larger"
+    },
+    csvLink: {
+      marginTop: "0.5rem"
+    },
+    newPageIcon: {
+      color: "white",
+      marginTop: "0.5rem",
+      width: "1rem",
+      height: "1rem"
+    },
+    loadMoreButton: {
+      margin: "auto",
+      display: "flex",
+      border: "none",
+      fontWeight: "bold",
+      marginTop: "1rem",
+      marginBottom: "1rem"
+    },
+    seenAllText: {
+      margin: "auto",
+      fontWeight: "bold",
+      marginTop: "1rem",
+      marginBottom: "1rem",
+      textAlign: "center"
     }
   })
 );
 
 interface Props {
+  sources: APISource[];
+  meta?: { num_found?: number };
   addConceptsToDictionary: (
     ...args: Parameters<typeof recursivelyAddConceptsToDictionaryAction>
   ) => void;
+  retrievePublicSources: (
+    ...args: Parameters<typeof retrievePublicSourcesAction>
+  ) => Promise<any>;
 }
 
-const AddBulkConceptsPage: React.FC<Props> = ({ addConceptsToDictionary }) => {
+const AddBulkConceptsPage: React.FC<Props> = ({ addConceptsToDictionary, sources = [], retrievePublicSources, meta = {} }) => {
   const classes = useStyles();
   const { pathname: url } = useLocation();
   const { fromSource } = useQueryParams();
+  const history = useHistory();
+
+  
+  const [showAllSources, setShowAllSources] = useState(false);
+
+  const [queryString, setQueryString] = useState('');
+  const handleSearch = (q: string) => setQueryString(q);
+
+  const [currentSources, setCurrentSources] = useState<
+    { name: string; sourceUrl: string }[]
+  >([]);
+
+  const defaultSources = Object.entries(
+    PREFERRED_SOURCES
+  ).map(([key, value]) => ({ name: key, sourceUrl: value }));
+
+  const allSources = defaultSources.concat(sources.map(s => ({ name: s.name, sourceUrl: s.url })))
+
+  const selectedSource = allSources?.find(s => s.name === fromSource);
+
+  useEffect(() => {
+    if (!selectedSource) {
+      const defaultQueryParam = 'fromSource=CIEL';
+      history.replace({
+        search: defaultQueryParam.toString(),
+      })
+    }
+  }, []); // eslint-disable-line
+
+  const conceptsUrl = `${selectedSource?.sourceUrl}concepts/`;
+
+  const [page, setPage] = useState<number>(1);
+  const [paginating, setPaginating] = useState(false);
+  
+  const { num_found: numberFound = sources.length } = meta;
+
+  const sourceUrl = "/sources/";
+  const limit = 20;
+
+  const loadSourcesList = () => {
+    setPaginating(true);
+    setPage(page+1);
+    retrievePublicSources(sourceUrl, queryString, limit, page)
```

Comment:
Quick thought here: for the full source list, this should actually be `retrievePrivateSources`. The `public` version will only retrieve publicly-visible sources. The `private` version will retrieve all sources including both private and public. Otherwise, we're restricting users to *only* load concepts from sources they've marked as `public`. We should allow users to load concepts from private sources, in case implementers don't wish to publicly publish their source.

Reply:
Thanks @ibacher for your review

Reply:
@ibacher this has been fixed ☝️